### PR TITLE
Wrap multiple VACOLS closures in a transaction

### PR DESCRIPTION
**Pull request by shanear**
Currently, if there are multiple appeals being closed in VACOLS, and one of them fails, it will not rollback the others. This is obviously a problem.

This PR creates `Appeal.close` which takes an an array of appeals and closes them all within a transaction, solving the issue.

It also allows a block to be provided which also will be run inside the VACOLS transaction. This will allow us to safely transactionalize the creation of the end product with the VACOLS appeal closings.

Wasn't sure a good way to do an automated test on the VACOLS rollback. I think it's going to be either very difficult or very janky and not worth doing. Went ahead and pushed it without it.

The happy path still works, but there isn't a great way to test the sad path, because you need a situation where one Appeal fails and the other succeeds. There will be a better way to test when we create the end product, since we can test that an end product creation failure rolls back the VACOLS closures.

connects #3749

### CODE REVIEW

File: app/models/appeal.rb [hunk 1]
```diff
@@ -534,18 +530,23 @@ def repository
...
+        (appeals || [appeal]).each do |appeal|
+          close_single(
+            appeal: appeal,
+            user: user,
+            closed_on: closed_on,
+            disposition: disposition
+          )
+        end
+
+        inside_transaction.call if block_given?
```
Comment (shanear) [inline app/models/appeal.rb]: This isn't used now, but will be when we start auto CESTing

File: app/services/appeal_repository.rb [hunk 2]
```diff
@@ -9,6 +9,12 @@ def self.vacols_db_connection_active?
     VACOLS::Record.connection.active?
   end
 
+  def self.transaction
```
Comment (shanear) [inline app/services/appeal_repository.rb]: I could have reached directly out to this method in `ramp_intake.rb` but decided to keep it encapsulated in `Appeal`

File: app/models/appeal.rb [hunk 3]
```diff
@@ -534,19 +530,26 @@ def repository
...
-
-      repository.close!(
-        appeal: appeal,
-        user: user,
-        closed_on: closed_on,
-        disposition_code: disposition_code
-      )
+    # Wraps the closure of appeals in a transaction
+    # add additional code inside the transaction by passing a block
+    # rubocop:disable Metrics/ParameterLists
```
Comment (shanear) [inline app/models/appeal.rb]: felt like the parameter list is the best way to do this for now, might wrap back around and see if I can't do something better.

File: lib/fakes/appeal_repository.rb [hunk 4]
```diff
@@ -1,6 +1,7 @@
 require "ostruct"
 
 # frozen_string_literal: true
+# rubocop:disable Metrics/ClassLength
```
Comment (shanear) [inline lib/fakes/appeal_repository.rb]: I added the 4 lines that broke the rubocops back 🐫

File: app/models/appeal.rb [hunk 5]
```diff
@@ -534,19 +530,26 @@ def repository
...
+    def close(appeal:nil, appeals:nil, user:, closed_on:, disposition:, &inside_transaction)
+      fail "Only pass either appeal or appeals" if appeal && appeals
```
Comment (nickheiner-usds) [inline app/models/appeal.rb]: 👍

File: app/models/appeal.rb [hunk 6]
```diff
@@ -578,6 +581,20 @@ def convert_file_number_to_vacols(file_number)
 
     private
 
+    def close_single(appeal:, user:, closed_on:, disposition:)
+      fail "Only active appeals can be closed" unless appeal.active?
+
+      disposition_code = VACOLS::Case::DISPOSITIONS.key(disposition)
+      fail "Disposition #{disposition}, does not exist" unless disposition_code
```
Comment (nickheiner-usds) [inline app/models/appeal.rb]: No need for the comma in the error message.

**Comment by nickheiner-usds** [comment]
This looks great. Thanks for thinking of the need to close all appeals transactionally and handling it.

I don't feel like I have enough context to be the sole reviewer on this. LGTM, but I would defer to @amprokop for final approval.

Have you tested this with a running VACOLS instance? This could be a good time to use the AWS VACOLS replica that @kierachell set up. :smile:

File: app/models/appeal.rb (hunk 1 above)
Comment (amprokop) [reply app/models/appeal.rb]: ahh this is a great way to call an optional block. thank you ruby master

**Comment by amprokop** [comment]
This is awesome, and great for safety. Have we experienced this actually happening before? Aside from if VACOLS goes down in the middle of closing several appeals, why would this fail?

LGTM anyway

File: app/models/appeal.rb (hunk 1 above)
Comment (shanear) [reply app/models/appeal.rb]: 🙇

**Comment by shanear** [comment]
No reasons I can think of, which makes this tough to test, but if there's anything I've learned its that things will always fail lol

**Comment by amprokop** [comment]
Lgtm

File: app/models/appeal.rb (hunk 3 above)
Comment (amprokop) [reply app/models/appeal.rb]: Cool

**Pull request merged**